### DISCUSSION TITLE
Compute_batch_utility from ConditionalLogit assumes the same item numbers as in train

### DISCUSSION THREAD

**Discussion by ghost** [artefactory/choice-learn#277] [General]
I was trying to simulate a test case scenario where I compute the utilities for a test ChoiceDataset.
<img width="779" height="762" alt="image" src="https://github.com/user-attachments/assets/6fc5f3b3-d65e-4120-977b-f69f511623ff" />
But then I encountered an error when trying to compute the utilities because of the number of trainable weights which was computed per training item it's different than the number of items that are in the test set:
<img width="798" height="286" alt="image" src="https://github.com/user-attachments/assets/dc949b1b-3007-421a-8fd0-9174301b8e8a" />
According to the documentation (2_data_handling.ipynb), 
_shared_features_by_choice  = (n_choices, n_shared_features)
items_features_by_choice = (n_choices, n_items, n_items_features) 
available_items_by_choice = (n_choices, n_items)_

Does the original implementation in ConditionalLogit.compute_batch_utility assumes to have the same number of items (**n_items**) for test as the number of items it was trained on? Why? 

I see that in the auxiliary_tools/assortment_example.ipynb, in the TaFengMNL class you overwrite the compute_batch_utility function and there the items_features_by_choice it's defined with shape (n_choices, n_items_features). Is it because of this reason to have the possibility to oscillate the number of items in the test set?

**Comment by VincentAuriau** [comment]
Hello,

Indeed the models assume that the number of items is the same and also with the same order.
You can have missing items in your test set but not in your train set.

I am not sure what happens here but what I can think of is that you have an item in your test set(297 for example) that is absent in your train set. The way the model works is with preset matrixes of weights, so for example compute_batch_utility needs to return a matrix with *(batch_shape, n_items)* shape with the indexes being the same as the choices values (from 0 to 295 if you have 296 items). The model is indeed not made to add a 297-th utility value in this matrix.

If you have a missing item in your train set then you cannot estimate any specific value for this item (for example an intercept). indeed, it will push the probability  of this item to be zero and therefore not interesting / right.
If you have only shared features it could theoretically work but in this case I am not sure if the code works since we haven't try it. If you need it I can test see if there are issues.

**Reply by ghost** [reply]
Than you! Yes this is exactly the scenario we've encountered - we do have some items from the test set which are not present in the train set, so therefore the utilities cannot be computed... 

Just to give you an idea of our problem formulation: We have transactions from last years (2018- august 2025) from which we compute a huge ChoiceDataset, and we want to predict the optimal assortment for season autumn/winter 2025. But then of course there would be cases where we have new items which are added for the new season and therefore those items are missing from the train set... 

One solution that came into my mind is that, we would include the past data (including last season autumn/winter 2024 which initially we thought could be preserved as a test set since this is the latest autumn/winter season we have information about), compute the utilities also on the same training set, and then "filter-out" the predicted assortment to preserve only autumn/winter items. I know it contradicts with the principles of a correct train-test split, but I don't see any other solution for this..

**Reply by VincentAuriau** [reply]
In this case, what I usually do is:
- A temporal split where I remove the additional items
- A store split 

The models might be different (no store effect in the second split for example) but it's the best solution I have found.

**Reply by ghost** [reply]
Okay understood, thank you! So this would solve the issue of the new items present in the test set and not in the train set. 

But still, how do we treat missing items in the test set but not in the train set (e.g. the n_items=296 for the train and n_items=57 in the test), since the utility matrix will be of (batch_shape, n_items)? Do we need to preserve the indexes of the items from test to match the items from train and when defining _items_features_by_choice_ in _compute_batch_utility_ we simply "zero-out" the indexes where we don't have a match?

**Reply by VincentAuriau** [reply]
Yes it is definitely important that the items indexes are aligned between train and test - otherwise the model has no way to be aware that items 24 might have swapped to item 2. It is similar as a classification problem, your classes need to be the same between train & test. (unless you do not have any item-specific weight, however I would still recommend to do it for the sake of "cleanliless").

What will happen with the test set is that it will output n_items utilities (=296), and compute the probabilities with the softmax. If some items aren't in the assortment their probability will be set to 0. Then the model will be evaluated with the probability outputed at index $x$ for $x$ the choice in the test set.

The *items_features_by_choice* will not matter so zeroing them is a good choice ! Just make sur that the *availability* matrix is well defined !

**Reply by ghost** [reply]
Got it! Thank you so much! ;)

**Comment by VincentAuriau** [comment]
Let me know if understood wrong - there also might be some bugs that need to be fixed :)
